src/vm/circuit.rs

```diff
@@ -0,0 +1,333 @@
+    };
+    let circ_info = match _registry.get_type(&_info.ty).unwrap() {
+        CoreTypeConcrete::Circuit(CircuitTypeConcrete::Circuit(info)) => &info.circuit_info,
+        _ => todo!(),
+    };
+    let mut outputs = vec![None; 1 + circ_info.n_inputs + circ_info.values.len()];
+    let mut add_gates = circ_info.add_offsets.iter().peekable();
+    let mut mul_gates = circ_info.mul_offsets.iter().skip(circ_info.n_inputs);
+
+    outputs[0] = Some(BigUint::from(1_u8));
+
+    for (i, input) in inputs.iter().enumerate() {
+        let gate = &circ_info.mul_offsets[i];
+        outputs[gate.output] = Some(input.to_owned());
+    }
+
+    let success = loop {
+        while let Some(add_gate) = add_gates.peek() {
+            let lhs = outputs[add_gate.lhs].to_owned();
+            let rhs = outputs[add_gate.rhs].to_owned();
+
+            match (lhs, rhs) {
+                (Some(l), Some(r)) => {
+                    outputs.insert(add_gate.output, Some((l + r) % &modulus));
+                }
+                (None, Some(r)) => {
+                    let res = match outputs[add_gate.output].to_owned() {
+                        Some(res) => res,
+                        None => break,
+                    };
+                    // if it is a sub_gate the output index is store in lhs
+                    outputs.insert(add_gate.lhs, Some((res + &modulus - r) % &modulus));
+                }
+                // there aren't enough gates computed for add_gate to compute
+                // the next gate so we need to compute a mul_gate
+                _ => break,
+            };
+
+            add_gates.next();
+        }
+
+        match mul_gates.next() {
+            Some(mul_gate) => {
+                let lhs = outputs[mul_gate.lhs].to_owned();
+                let rhs = outputs[mul_gate.rhs].to_owned();
+
+                match (lhs, rhs) {
+                    (Some(l), Some(r)) => {
+                        outputs.insert(mul_gate.output, Some((l * r) % &modulus));
+                    }
+                    (None, Some(r)) => {
+                        let res = match r.modinv(&modulus) {
+                            Some(inv) => inv,
+                            None => {
+                                // attempted to get the inverse of 0,
+                                break false;
+                            }
+                        };
+                        // if it is a inv_gate the output index is store in lhs
+                        outputs.insert(mul_gate.lhs, Some(res));
+                    }
+                    // this state should be reached since it would mean that
+                    // not all the circuit's inputs where filled
+                    _ => unreachable!(),
+                }
+            }
+            None => break true,
+        }
+    };
+
+    if success {
+        EvalAction::NormalBranch(
+            0,
+            smallvec![add_mod, mul_mod, Value::CircuitOutputs(outputs),],
+        )
```

Comment:
In case of success, all outputs will be defined, that is, no element will have None value. With that in mind, you can make the following optimizations (so that we have the same representation of Cairo Native):
- The first `1 + n_values` can be skipped, as they correspond to input data and not to any gate
- You can remove the Option and have a `Vec<BigUint>`
- You should use indexing rather than calling insert.

I tested a simple add circuit. This is the CircuitOutput I got:
```
"CircuitOutputs": [
  [
    1
  ],
  null,
  null,
  [
    3
  ],
  [
    6
  ],
  [
    2
  ],
  null
]
```
To mimic Cairo Native, it should look like this:
```
"CircuitOutputs": [
  [
    3
  ],
  [
    6
  ],
  [
    2
  ],
]
```